File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveEnumMemberAccess.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveEnumMemberAccess;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\References\EnumMemberReference;
use OneLang\One\Ast\References\EnumReference;

class ResolveEnumMemberAccess extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveEnumMemberAccess");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof PropertyAccessExpression && $expr->object instanceof EnumReference;
    }
    
    function transform($expr) {
        $pa = $expr;
        $enumMemberRef = $pa->object;
        $member = \OneCore\ArrayHelper::find($enumMemberRef->decl->values, function ($x) use ($pa) { return $x->name === $pa->propertyName; });
        if ($member === null) {
            $this->errorMan->throw("Enum member was not found: " . $enumMemberRef->decl->name . "::" . $pa->propertyName);
            return $expr;
        }
        return new EnumMemberReference($member);
    }
    
    function canDetectType($expr) {
        return $expr instanceof EnumMemberReference;
    }
    
    function detectType($expr) {
        $expr->setActualType(($expr)->decl->parentEnum->type);
        return true;
    }
}

File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveEnumMemberAccess.py
from onelang_core import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs

class ResolveEnumMemberAccess(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveEnumMemberAccess")
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.PropertyAccessExpression) and isinstance(expr.object, refs.EnumReference)
    
    def transform(self, expr):
        pa = expr
        enum_member_ref = pa.object
        member = next(filter(lambda x: x.name == pa.property_name, enum_member_ref.decl.values), None)
        if member == None:
            self.error_man.throw(f'''Enum member was not found: {enum_member_ref.decl.name}::{pa.property_name}''')
            return expr
        return refs.EnumMemberReference(member)
    
    def can_detect_type(self, expr):
        return isinstance(expr, refs.EnumMemberReference)
    
    def detect_type(self, expr):
        expr.set_actual_type((expr).decl.parent_enum.type)
        return True